<a href="https://colab.research.google.com/github/dori108/gdg-study/blob/main/4_1_%EB%A0%88%EC%A6%88%EB%84%B7_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1. 사전 학습된 모델의 성능 체험
ResNet50(weights='imagenet')은 ImageNet 데이터셋에서 이미 훈련된 모델
→ 복잡한 모델을 직접 학습하지 않아도 고성능 분류를 즉시 테스트할 수 있음

2. 이미지 분류 결과 확인
'4-7-푸들.png' 이미지를 입력하고,

ResNet50이 이 이미지를 어떻게 인식하는지를 Top-3 예측 결과로 출력
→ 예: 'poodle' 확률: 95.2%처럼 결과가 나옴

3. 이미지 전처리와 추론 과정 학습
image.load_img(), preprocess_input(), tf.expand_dims() 등
→ 이미지 분류 입력 파이프라인 흐름을 익히자!

In [ ]:
!gdown 1WI3FPChY3DFnrNEP3Fk_C61bcMxkr4-b
import tensorflow as tf
from tensorflow.keras.applications import ResNet50
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.resnet import preprocess_input, decode_predictions

# ResNet 모델 불러오기
resnet_model = ResNet50(weights='imagenet', include_top=True)

# 이미지 전처리 및 예측
PATH = './'
FILE_NAME = '4-7-푸들.png'
img = image.load_img(PATH + FILE_NAME, target_size=(224, 224))
input_image = image.img_to_array(img)
input_image = preprocess_input(input_image)
input_image = tf.expand_dims(input_image, axis=0)

predictions = resnet_model.predict(input_image)
decoded_predictions = decode_predictions(predictions, top=3)[0]

for class_id, class_name, probability in decoded_predictions:
  print(f"{class_name} 확률: {probability:.2f}%")



Downloading...
From: https://drive.google.com/uc?id=1WI3FPChY3DFnrNEP3Fk_C61bcMxkr4-b
To: /content/4-7-푸들.png
100% 113k/113k [00:00<00:00, 67.3MB/s]
102967424/102967424 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 3s 3s/step
35363/35363 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
miniature_poodle 확률: 0.66%
toy_poodle 확률: 0.30%
standard_poodle 확률: 0.02%


왜 이렇게 낮은 확률이 나왔을까?
입력 이미지가 ResNet50이 기대한 품질과 다를 수 있음
흐리거나, 너무 가까이 찍었거나, 배경이 복잡할 수도 있음
이미지 사이즈나 채널 정규화 오류

preprocess_input()을 안 썼거나, float32로 변환되지 않은 경우엔 확률이 낮게 나올 수 있어요.
→ 하지만 이 코드에서는 전처리를 잘 했기 때문에 가능성은 낮습니다.

ResNet50은 ImageNet 기준으로 훈련됨

ImageNet에 있는 푸들 사진과 여러분의 사진이 스타일이나 구도가 너무 다르면
→ 자신 있게 예측하지 못할 수 있습니다.

출력값이 Softmax이기 때문에 전체 클래스 중 상대적인 확률

1000개의 클래스 중에 가장 높은 확률이 **0.66%**인 것일 수 있어요.
→ 이 경우 모델이 자신 없는 상태라고 볼 수 있습니다.